contracts/btc-staking/src/staking.rs

```diff
@@ -322,6 +398,164 @@ pub fn handle_slash_fp(
     slash_finality_provider(deps, env, fp_btc_pk_hex)
 }
 
+pub fn handle_distribute_rewards(
+    mut deps: DepsMut,
+    env: &Env,
+    info: &MessageInfo,
+    rewards: &[RewardInfo],
+) -> Result<Vec<Event>, ContractError> {
+    let params = PARAMS.load(deps.storage)?;
+
+    // TODO: Check that the sender is the finality contract (AML)
+
+    // Error if no proper funds to distribute
+    let amount = must_pay(info, &params.rewards_denom)?;
+
+    // Check that the total rewards match sent funds
+    let total_amount: Uint128 = rewards.iter().map(|r| r.reward).sum();
+    if total_amount != amount {
+        return Err(ContractError::InvalidRewardsAmount(amount, total_amount));
+    }
+
+    rewards
+        .iter()
+        .map(|reward_info| {
+            distribute_rewards(
+                &mut deps,
+                &reward_info.fp_pubkey_hex,
+                reward_info.reward,
+                env.block.height,
+            )
+        })
+        .collect()
+}
+
+fn distribute_rewards(
```

Comment:
Does this function distributes commission to FP's address, or this happens somewhere else?

Reply:
Commissions are not implemented yet. I suggest we do it in a follow-up.